### PR TITLE
Add actions tasks

### PR DESCRIPTION
**Pull request by imbstack**
This is an implementation of the [action tasks stuff](http://gecko.readthedocs.io/en/latest/taskcluster/taskcluster/actions.html) in the tools site.

Forgive me for however I have sinned against React. This _works_ when I run it locally, but I make no other guarantees 🔨 .

There are some nice improvements that can be made later UX-wise, but I wanted to get the basics in with a not-huge review.

### CODE REVIEW

File: src/components/Error/index.js [hunk 1]
```diff
@@ -34,19 +43,27 @@ export default class Error extends React.PureComponent {
         <strong>{this.getTitle()}&nbsp;</strong>
         <Markdown>
           {(() => {
-            if (!error.message || process.env.NODE_ENV === 'development') {
+            if (!error.message) {
```
Comment (djmitche) [inline src/components/Error/index.js]: I think we want to keep this so we get stacks in dev.

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 2]
```diff
@@ -240,19 +266,24 @@ export default class ActionsMenu extends React.PureComponent {
...
-
-    if (!taskId) {
-      return <NavItem disabled>Task Actions</NavItem>;
-    }
+  taskInContext(tagSetList, taskTags) {
+    return tagSetList.some(tagSet =>
+      Object.keys(tagSet).every(tag => (
+        taskTags[tag] && taskTags[tag] === tagSet[tag]
+      ))
+    );
```
Comment (djmitche) [inline src/views/UnifiedInspector/ActionsMenu.js]: could Set make this more efficient?

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 3]
```diff
@@ -308,6 +339,36 @@ export default class ActionsMenu extends React.PureComponent {
           body={this.editInteractiveModal()}>
           <Icon name="edit" /> Edit as Interactive Task
         </ModalItem>
+
+        <MenuItem divider />
+        <MenuItem header>Task Actions</MenuItem>
+        { taskActions && taskActions.map((action, i) => (
```
Comment (djmitche) [inline src/views/UnifiedInspector/ActionsMenu.js]: maybe a disabled "loading.." item here when loading?

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 4]
```diff
@@ -308,6 +339,36 @@ export default class ActionsMenu extends React.PureComponent {
...
+        <MenuItem divider />
+        <MenuItem header>Group Actions</MenuItem>
+        { groupActions && groupActions.map((action, i) => (
+          <ActionModal
+            action={action}
+            credentials={credentials}
+            decision={decision}
+            actionVariables={actions.variables}
+            taskGroupId={taskGroupId}
+            key={i}>
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Keys should not be plain iteration indices as they could conflict with other React keys. Switch this to something like:

```jsx
key={`action-modal-variable-${i}`}>
```

File: src/views/UnifiedInspector/Inspector.js [hunk 5]
```diff
@@ -408,8 +434,11 @@ export default class Inspector extends React.PureComponent {
   }
 
   renderTaskGroup() {
-    const { taskGroupId, taskId, queue, purgeCache, url, runId, sectionId, subSectionId, artifactId } = this.props;
-    const { tasks, status, task, selectedTaskId, selectedRun, artifacts } = this.state;
+    const { taskGroupId, taskId, queue, purgeCache,
```
Comment (eliperelman) [inline src/views/UnifiedInspector/Inspector.js]: If this doesn't fit on 1 line, put each variable on its own line.

File: src/views/UnifiedInspector/Inspector.js [hunk 6]
```diff
@@ -103,6 +107,28 @@ export default class Inspector extends React.PureComponent {
...
+  async loadActions(props) {
+    const { queue, taskGroupId } = props;
+
+    if (!taskGroupId) {
+      return;
+    }
+
+    try {
+      const decision = await queue.task(taskGroupId);
+
```
Comment (eliperelman) [inline src/views/UnifiedInspector/Inspector.js]: Remove extra blank line.

File: src/components/ActionModal/index.js [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+              <h4>Schema</h4>
+              <Code language="yaml" style={{ maxHeight: 250, overflow: 'scroll' }}>
+                {safeDump(action.schema || {})}
+              </Code>
+            </Col>
+          </Row>
+          <br />
+          {error && <Error error={error} />}
+          {validation && validation !== 'success' &&
+            <Alert bsStyle="danger">
```
Comment (eliperelman) [inline src/components/ActionModal/index.js]: Wrap multi-line JSX in parens:

```jsx
{validation && validation !== 'success' && (
  <Alert bsStyle="danger">
    <strong>Validation Error</strong>
    <pre>{validation}</pre>
  </Alert>
)}
```

File: package.json [hunk 8]
```diff
@@ -31,6 +32,8 @@
     "hterm-umd": "^1.0.1",
     "jquery": "^3.1.0",
     "js-yaml": "^3.7.0",
+    "json-e": "^2.1.1",
+    "json-schema-defaults": "^0.3.0",
```
Comment (jonasfj) [inline package.json]: yeah, ajv has an option for this...

Comment (imbstack) [inline package.json]: We've cleared up this confusion and are going to stick with using both ajv and json-schema-defaults.

File: src/components/Error/index.js (hunk 1 above)
Comment (imbstack) [inline src/components/Error/index.js]: Makes sense. It did hide the `error.message` in development though. I'm going to make a compromise in the next revision that should satisfy both needs.

File: src/views/UnifiedInspector/ActionsMenu.js (hunk 2 above)
Comment (imbstack) [inline src/views/UnifiedInspector/ActionsMenu.js]: I'm not sure it could, but I might be missing something. What are you thinking of?

Comment (djmitche) [inline src/views/UnifiedInspector/ActionsMenu.js]: It's probably nothing -- just that set intersection might be a faster operation.

**Comment by imbstack** [comment]
Ok, I think this works without a new component now!

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 9]
```diff
@@ -240,19 +306,86 @@ export default class ActionsMenu extends React.PureComponent {
...
+        taskTags[tag] && taskTags[tag] === tagSet[tag]
+      ))
+    );
+  }
+
+  actionTaskModal = (name) => {
+    const { actionInputs, actionData } = this.state;
+    const { action } = actionData[name];
+    const form = actionInputs[name];
+
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Remove extra blank line.

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 10]
```diff
@@ -240,19 +306,86 @@ export default class ActionsMenu extends React.PureComponent {
...
+    const handleFormChange = (value) => {
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Move this to a class method.

```js
handleFormChange = (value, name) => this.setState({
  actionInputs: {
    ...this.state.actionInputs,
    [name]: value
  }
});

actionTaskModal = (name) => {
  // ...
  <CodeEditor mode="yaml" lint={true} value={form} onChange={value => this.handleFormChange(value, name)} />
```

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 11]
```diff
@@ -240,19 +306,86 @@ export default class ActionsMenu extends React.PureComponent {
...
+    );
+  }
+
+  actionTaskSubmit = name => async () => {
+    const { task, taskId, taskGroupId, actions } = this.props;
+    const { actionInputs, actionData } = this.state;
+    const form = actionInputs[name];
+    const { validate, action } = actionData[name];
+    const input = safeLoad(form);
+
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Remove extra blank line.

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 12]
```diff
@@ -240,19 +306,86 @@ export default class ActionsMenu extends React.PureComponent {
...
+    const valid = validate(input);
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Insert blank line between declarations and other statements.

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 13]
```diff
@@ -240,19 +306,86 @@ export default class ActionsMenu extends React.PureComponent {
...
 
-    const { caches } = this.state;
-    const { queue, purgeCache, status, task, onRetrigger, onEdit, onCreateInteractive, onEditInteractive } = this.props;
+    const newTask = jsone(action.task, merge({
+      taskGroupId,
+      taskId,
+      task,
+      input
+    }, actions.variables));
+
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Remove extra blank line.

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 14]
```diff
@@ -240,19 +306,86 @@ export default class ActionsMenu extends React.PureComponent {
...
+    const newTaskId = nice();
+    await this.actionsQueue.createTask(newTaskId, newTask);
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Insert blank line above this `await`.

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 15]
```diff
@@ -240,19 +306,86 @@ export default class ActionsMenu extends React.PureComponent {
...
+    return newTaskId;
+  }
+
+  render() {
+    const { caches, actionData, taskActions, groupActions } = this.state;
+    const { queue,
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Move alignment to two spaces, starting with the first variable on its own line:

```js
const {
  queue,
  purgeCache,
  ...
} = this.props;
```

File: src/views/UnifiedInspector/Inspector.js [hunk 16]
```diff
@@ -408,8 +435,26 @@ export default class Inspector extends React.PureComponent {
   }
 
   renderTaskGroup() {
-    const { taskGroupId, taskId, queue, purgeCache, url, runId, sectionId, subSectionId, artifactId } = this.props;
-    const { tasks, status, task, selectedTaskId, selectedRun, artifacts } = this.state;
+    const { taskGroupId,
```
Comment (eliperelman) [inline src/views/UnifiedInspector/Inspector.js]: Move alignment to two spaces, starting with the first variable on its own line:

```js
const {
  taskGroupId,
  taskId,
  ...
} = this.props;
const {
  tasks,
  actions,
  ...
} = this.state;
```

File: src/views/UnifiedInspector/Inspector.js [hunk 17]
```diff
@@ -408,8 +435,27 @@ export default class Inspector extends React.PureComponent {
...
+      queue,
+      purgeCache,
+      url,
+      runId,
+      sectionId,
+      subSectionId,
+      artifactId,
+      credentials
+    } = this.props;
+    const { tasks,
```
Comment (eliperelman) [inline src/views/UnifiedInspector/Inspector.js]: Spacing here:

```js
const {
  tasks,
  actions,
  // ...
} = this.state;
```

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 18]
```diff
@@ -43,6 +59,56 @@ export default class ActionsMenu extends React.PureComponent {
...
+      const groupActions = [];
+      const actionInputs = {};
+      const actionData = {};
+
+      if (actions && actions.actions) {
+        actions.actions.forEach((action) => {
+          const schema = action.schema || {};
+          const validate = this.ajv.compile(schema);
+          const form = safeDump(jsonSchemaDefaults(schema) || {});
+          actionInputs[action.name] = form;
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Insert blank line.

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 19]
```diff
@@ -43,6 +59,56 @@ export default class ActionsMenu extends React.PureComponent {
...
+          actionData[action.name] = {
+            action,
+            validate
+          };
+          if (!action.context.length) {
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: Insert blank line above this conditional.

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 20]
```diff
@@ -308,6 +443,38 @@ export default class ActionsMenu extends React.PureComponent {
...
+          <ModalItem
+            modalSize={actionData[action].action.schema ? 'large' : null}
+            body={this.actionTaskModal(action)}
+            onSubmit={this.actionTaskSubmit(action)}
+            onComplete={onActionTask}
+            key={`taskaction-modal-item-${action}`}>
+            <Icon name="keyboard-o" /> {actionData[action].action.title}
+          </ModalItem>
+        ))}
+        {taskActions === null && (
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: You can 1-line this and remove the interpolation for string tag contents:

```jsx
{taskActions === null && <MenuItem disabled={true}>Loading...</MenuItem>}
```

File: src/views/UnifiedInspector/ActionsMenu.js [hunk 21]
```diff
@@ -308,6 +443,38 @@ export default class ActionsMenu extends React.PureComponent {
...
+          <ModalItem
+            modalSize={actionData[action].action.schema ? 'large' : null}
+            body={this.actionTaskModal(action)}
+            onSubmit={this.actionTaskSubmit(action)}
+            onComplete={onActionTask}
+            key={`groupaction-modal-item-${action}`}>
+            <Icon name="keyboard-o" /> {actionData[action].action.title}
+          </ModalItem>
+        ))}
+        {groupActions === null && (
```
Comment (eliperelman) [inline src/views/UnifiedInspector/ActionsMenu.js]: You can 1-line this and remove the interpolation for string tag contents:

```jsx
{groupActions === null && <MenuItem disabled={true}>Loading...</MenuItem>}
```

**Pull request merged**